### PR TITLE
handle redirection on ssr 4647

### LINKED ISSUE #4647: handle redirection on client and on server
## What is the motivation for adding / enhancing this feature?

Right now if we want to redirect user we need to check if he is on client or server side. In first case we use router and in second we need to get access to server context and then we can make redirection in express app.
For example;
```
      const newUrl = '/women/tops-women/tees-women/tees-25/karissa-v-neck-tee-1529.html?childSku=WS10-XS-Green'
      AsyncDataLoader.push({
        execute: async ({ context }) => {
          if (context && !context.url.includes(newUrl)) {
            context.server.response.redirect(301, newUrl)
          }
        }
      })
```
or same thing in `asyncData`.
I think this is a common problem which should be resolved by service or helper in core.

Also right now when we make redirection on server we can get error `Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client`. That is because in server.ts we still call res.send().

Last thing is that we repeat data fetch in each request. I don't know if this can be solved.
For example:
Lets say that user goes to product page then he needs to fetch product, menu, breadcrumbs (if menu is lazy loaded), attributes...
Now at the end of that we have enough data to decide if he needs to be redirected to other page. After redirection it goes once again => fetch product, menu...

If we have redis then maybe we can add vuex data from redirected request to redis. After redirection we can check if there is data in redis (and remove it, this data should exist only for next request) and we can add it as initial vuex state. In that case we don't need to make requests again.

## What are the acceptance criteria 

- [ ] add helper or service which will handle redirection on ssr and csr
- [ ] handle redirection in server.ts after `renderer.renderToString(context)` is finished
- [ ] (optional) keep vuex state between requests

## Can you complete this feature request by yourself?

- [x] YES
- [ ] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [x] This is a normal feature request. This should be available on https://test.storefrontcloud.io and then after tests this can be added to next Vue Storefront version. In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] (**Pick this option only if you're sure**) This is an important improvement request for current Release Candidate version on https://next.storefrontcloud.io and should be placed in next RC version. In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] (**Pick this option only if you're sure**) This is a critical improvement request for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version. In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

### PR DESCRIPTION
**Pull request by gibkigonzo**
### Related Issues

closes #4647
related #4903 - we can use redirect here

### Short Description and Why It's Useful

 [x] - add helper or service which will handle redirection on ssr and **csr** (I didn't handle it on client side, because I think there is no need to. It's easy to be done and there are no problems with it. We can use router.push/router.replace in any place)
 [x] - handle redirection in server.ts after renderer.renderToString(context) is finished
 [ ] - (optional) keep vuex state between requests - TODO

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [x] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/scripts/server.ts [hunk 1]
```diff
@@ -194,6 +194,11 @@ app.get('*', (req, res, next) => {
     }
     const context = ssr.initSSRRequestContext(app, req, res, config)
     renderer.renderToString(context).then(output => {
+      if (context.server._redirect.isPending()) {
```
Comment (Fifciu) [inline core/scripts/server.ts:197]: It looks good. However, what would happen if we want to redirect the user from the cached page? I think we should also support this case

File: core/scripts/utils/ssr-redirect.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  // lock path
+  if (typeof path !== 'string') throw new Error(`Redirection path needs to be a string`)
+  this.server._redirect.pendingPath = path
+
+  this.server._redirect.resolve = resolve.bind(this, path, options || {})
+}
+
+export function createContextRedirection (): SsrRedirect {
+  this.server._redirect = {
+    isPending: isPending.bind(this)
```
Comment (Fifciu) [inline core/scripts/utils/ssr-redirect.ts:47]: In this moment `this.server._redirect` still equals `null` right? If so wouldn't `return !!this.server._redirect.pendingPath` throw `Error: pendingPath of null`?

File: core/scripts/utils/ssr-redirect.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/*
+ * Resolve redirection only once per request.
+ */
+function resolve (path: string, {
+  code = 302
+}: RedirectOptions) {
+  // normal express redirection
+  this.server.response.redirect(code, path)
+  // clear temp object
+  Object.keys(this.server._redirect).forEach((key) => delete this.server._redirect[key])
```
Comment (Fifciu) [inline core/scripts/utils/ssr-redirect.ts:21]: Couldn't we just use?:
```js
this.server._redirect = {};
```

**Review by Fifciu — CHANGES_REQUESTED** [review]
Please check my suggestions/questions. I also recommend to write:
- Simple doc - how to redirect with asyncData/AsyncDataLoader
- Unit tests for `ssr-redirect.ts`

File: core/scripts/server.ts (hunk 1 above)
Comment (gibkigonzo) [reply core/scripts/server.ts:198]: In that case you need to clear cache. `dynamicRequestHandler` will not be resolved for cached pages. So if you changed urls, you need to clear cache and apply logic in asyncData that will handle redirection. I don't see other way :thinking:

File: core/scripts/utils/ssr-redirect.ts (hunk 2 above)
Comment (gibkigonzo) [reply core/scripts/utils/ssr-redirect.ts]: `this.server._redirect.pendingPath` it's not possible to call isPending before context is created - but it's outdated

File: core/scripts/utils/ssr-redirect.ts (hunk 3 above)
Comment (gibkigonzo) [reply core/scripts/utils/ssr-redirect.ts]: I used assign to null - outdated

**Comment by gibkigonzo** [comment]
**update** - added proxy for express reponse object to match api
Todo: tests + docs

File: core/scripts/server.ts (hunk 1 above)
Comment (Fifciu) [reply core/scripts/server.ts:198]: Couldn't we apply a similar mechanism to dynamicCacheHandler?

**Comment by gibkigonzo** [comment]
`[ ] - (optional) keep vuex state between requests ` - we can't initialize state with values, because after store initialization modules override state. So it can't be done

File: core/scripts/utils/ssr-redirect.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  if (!context.server._redirect) {
+    console.warn(`Something went wrong while creating vsf redirection handler. Fallback to normal express redirect`)
+    return expressResponse
+  }
+
+  const ProxyConstructor = Proxy || require('proxy-polyfill/src/proxy')
+
+  // returns proxy for Express Reponse object
+  return new ProxyConstructor(expressResponse, {
+    get (target, propKey) {
```
Comment (Fifciu) [inline core/scripts/utils/ssr-redirect.ts:87]: Couldn't we do it like that:
```js
get (target, propKey) {
      // 'redirect' can be called multiple times and we want to reuse same method
      if (propKey === 'redirect') {
        return context.server._redirect.handler
      }

      const originalMethod = target[propKey]
      return originalMethod;
    }
```

File: core/scripts/utils/types/index.ts [hunk 5]
```diff
@@ -12,12 +12,45 @@ export interface Context {
...
+   * it will play nicely with mounted apps, relative paths,
+   * `"back"` etc.
+   *
+   * Examples:
+   *
+   *    res.redirect('/foo/bar');
+   *    res.redirect('http://example.com');
+   *    res.redirect(301, 'http://example.com');
+   *    res.redirect('http://example.com', 301);
+   *    res.redirect('../login'); // /blog/post/1 -> /blog/login
```
Comment (Fifciu) [inline core/scripts/utils/types/index.ts:51]: Just wanted to say, it is so cool case. Glad it is supported!

**Review by Fifciu — CHANGES_REQUESTED** [review]

File: core/scripts/utils/ssr-redirect.ts [hunk 6]
```diff
@@ -85,13 +85,12 @@ export function createRedirectProxy (context: Context, expressResponse: Response
   // returns proxy for Express Reponse object
```
Comment (Fifciu) [inline core/scripts/utils/ssr-redirect.ts:85]: Reponse in comment! Fix it if you want, PR looks great right now.

**Review by Fifciu — APPROVED** [review]

File: core/scripts/utils/ssr-redirect.ts (hunk 6 above)
Comment (gibkigonzo) [reply core/scripts/utils/ssr-redirect.ts]: done

**Pull request merged**